File: backend/app/schemas/kg.py
"""
Knowledge Graph (KG) API Pydantic schemas (P2 T1).

- /api/v1/kg/graph        : 노드/엣지/통계
- /api/v1/kg/node/{id}/samples : 노드별 원문 샘플 (VOC)
- /api/v1/kg/search       : 텍스트 검색 (product/platform/category/country/keyword)

source: kg_edges_daily MV (P2-1 산출) — source/target 는 'type:label' 포맷.
"""
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


# ── /kg/graph ────────────────────────────────────────────────

class KGNode(BaseModel):
    id: str                         # 'product:GS25' 등
    type: str                       # 'product' | 'category' | 'platform' | 'country'
    label: str                      # 표시용 라벨 (product 는 name_ko/en, 그 외는 코드)
    count: int                      # 노드가 등장한 엣지 weight 합계
    sent_avg: float                 # 가중 평균 sentiment_score


class KGEdge(BaseModel):
    source: str
    target: str
    type: str                       # edge_type
    weight: int
    sent_avg: float


class KGStats(BaseModel):
    nodes_count: int
    edges_count: int
    period: Dict[str, str]          # {'start': 'YYYY-MM-DD', 'end': 'YYYY-MM-DD'}


class KGGraphResponse(BaseModel):
    nodes: List[KGNode]
    edges: List[KGEdge]
    stats: KGStats


# ── /kg/node/{id}/samples ───────────────────────────────────

class KGNodeSample(BaseModel):
    voc_id: int
    snippet: str
    sentiment_label: Optional[str] = None
    sentiment_score: Optional[float] = None
    source_url: Optional[str] = None
    platform_code: Optional[str] = None
    country_code: Optional[str] = None
    published_at: Optional[str] = None


# ── /kg/search ───────────────────────────────────────────────

class KGSearchHit(BaseModel):
    type: str                       # 'product' | 'platform' | 'category' | 'keyword'
    id: str                         # type:label 또는 keyword 자체
    label: str
    score: float                    # 단순 랭킹 점수 (등장빈도 또는 일치 강도)

File: frontend/src/types/kg.ts
// 지식 그래프(Knowledge Graph) 타입 정의 — P2-3 T1
//
// 백엔드 /api/v1/kg/* 응답 스키마와 1:1 대응.
// - 노드: id = "type:code" (예: "product:GS25U", "category:battery", "platform:reddit", "country:KR")
// - 엣지: (source, target, weight, sent_avg, edge_type)
// kg_edges_daily MV 의 product↔category / product↔platform / product↔country 3가지 edge_type 지원.

export type KGNodeType = 'product' | 'category' | 'platform' | 'country';

export type KGEdgeType = 'product_category' | 'product_platform' | 'product_country';

export interface KGNode {
  id: string;          // 예: "product:GS25U"
  label: string;       // 표시명
  type: KGNodeType;
  count: number;       // 누적 weight 또는 등장 횟수
  sent_avg: number | null; // -1.0 ~ +1.0 평균 감성
}

export interface KGEdge {
  source: string;
  target: string;
  weight: number;
  sent_avg: number | null;
  edge_type: KGEdgeType;
}

export interface KGGraphResponse {
  nodes: KGNode[];
  edges: KGEdge[];
}

export interface KGSample {
  voc_id: number;
  title: string | null;
  excerpt: string;
  platform_code: string | null;
  country_code: string | null;
  sentiment_score: number | null;
  url: string | null;
  collected_at: string; // ISO
}

export interface KGNodeSamplesResponse {
  node_id: string;
  samples: KGSample[];
}

export interface KGSearchHit {
  id: string;
  label: string;
  type: KGNodeType;
  count: number;
}

export interface KGSearchResponse {
  hits: KGSearchHit[];
}

// 컨트롤 패널 상태
export interface KGControls {
  topN: number;          // 노드 상위 개수 (40 ~ 200)
  minWeight: number;     // 엣지 최소 weight (1 ~ 50)
  edgeTypes: KGEdgeType[];
}

export const DEFAULT_KG_CONTROLS: KGControls = {
  topN: 80,
  minWeight: 3,
  edgeTypes: ['product_category', 'product_platform', 'product_country'],
};

// 노드 type → 색상 매핑 (요구사항)
export const NODE_TYPE_COLOR: Record<KGNodeType, string> = {
  product: '#1677ff',   // blue
  category: '#52c41a',  // green
  platform: '#fa8c16',  // orange
  country: '#722ed1',   // purple
};
